### PR TITLE
MOTECH-2442: add removing comments from xml template

### PR DESCRIPTION
**Pull request by matkwiatkowski**

### CODE REVIEW

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelper.java [hunk 1]
```diff
@@ -73,7 +75,12 @@ public void handleAction(Map<String, Object> parameters) throws IOException, Cla
         StreamingTemplateEngine streamingTemplateEngine = new StreamingTemplateEngine();
         Template template = streamingTemplateEngine.createTemplate(new String(ArrayUtils.toPrimitive(templateData)));
         Writable writable = template.make(parameters);
-        LOGGER.info("Template with name {}:\n{}", cdaTemplate.getTemplateName(), writable.toString());
-        iheTemplateService.sendTemplateToRecipientUrl(hl7Recipient.getRecipientUrl(), writable.toString());
+        String result = writable.toString();
+        Pattern regex = Pattern.compile("(?s)<!--.*?-->", Pattern.DOTALL);
```
Comment (pgesek) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelper.java]: Won't this be a performance suicide given how large our templates are? Wouldn't it make more sense to handle this during save?

**Comment by kmadej** [comment]
can you check if this new template engine supports removing comments ?

**Comment by matkwiatkowski** [comment]
I've checked groovy docs and there is no support for removing comments in xml for StreamingTemplateEngine

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java [hunk 2]
```diff
@@ -27,7 +29,25 @@ public CdaTemplate() {
     public CdaTemplate(String templateName, Byte[] templateData, Map<String, String> properties) {
         this.templateName = templateName;
         if (templateData != null) {
-            this.templateData = templateData.clone();
+
+            byte[] bytes = new byte[templateData.length];
```
Comment (kmadej) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java]: are you sure this works as expected ? As far as I know the CdaTemplate object is added via MDS, I'm not sure, however probably the new empty instance is created and then values are set separately, so this constructor will not be used. Generally adding this kind of parsing in constructors it's not the best idea

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java [hunk 3]
```diff
@@ -57,4 +59,26 @@ public void setTemplateData(Byte[] templateData) {
     public void setProperties(Map<String, String> properties) {
         this.properties = properties;
     }
+
+    private Byte[] removeCommentsFromTemplate(Byte[] data) {
+
+        byte[] bytes = new byte[data.length];
+        int i = 0;
+        for (Byte b : data) {
+            bytes[i++] = b.byteValue();
```
Comment (kmadej) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java]: byteValue() is redundant

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java [hunk 4]
```diff
@@ -57,4 +59,26 @@ public void setTemplateData(Byte[] templateData) {
...
+        }
+        String templateString = new String(bytes);
+        Pattern regex = Pattern.compile("(?s)<!--.*?-->", Pattern.DOTALL);
+        Matcher regexMatcher = regex.matcher(templateString);
+        templateString = regexMatcher.replaceAll("");
+        templateString.replaceAll("(?s)<!--.*?-->", "");
```
Comment (kmadej) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java]: this line is also redundant

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java [hunk 5]
```diff
@@ -57,4 +59,26 @@ public void setTemplateData(Byte[] templateData) {
...
+        bytes = templateString.getBytes();
+        Byte[] byteObjects = new Byte[bytes.length];
+        i = 0;
+        for (byte b : bytes) {
+            byteObjects[i++] = new Byte(b);
```
Comment (kmadej) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java]: using constructor redundant (autoboxing mechanism)

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java [hunk 6]
```diff
@@ -57,4 +59,26 @@ public void setTemplateData(Byte[] templateData) {
     public void setProperties(Map<String, String> properties) {
         this.properties = properties;
     }
+
+    private Byte[] removeCommentsFromTemplate(Byte[] data) {
+
```
Comment (kmadej) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/CdaTemplate.java]: empty line

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    public void shouldRemoveCommentsFromXml() {
+
+        String templateWithComments = "<ClinicalDocument>\n" +
+                "<!--  This is comment line -->\n" +
+                "</ClinicalDocument>";
+        byte[] bytes = templateWithComments.getBytes();
+        Byte[] byteObjects = new Byte[bytes.length];
+        int i = 0;
+        for (byte b : bytes) {
+            byteObjects[i++] = new Byte(b);
```
Comment (kmadej) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java]: autoboxing

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        String templateWithoutComments = "<ClinicalDocument>\n" +
+                "\n" +
+                "</ClinicalDocument>";
+        Map<String, String> properties = new HashMap<>();
```
Comment (kmadej) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java]: unused

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+        CdaTemplate cdaTemplate = new CdaTemplate();
+        cdaTemplate.setTemplateData(byteObjects);
+        
+        bytes = new byte[cdaTemplate.getTemplateData().length];
+        i = 0;
+        for (Byte b : cdaTemplate.getTemplateData()) {
+            bytes[i++] = b.byteValue();
```
Comment (kmadej) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java]: autoboxing

File: ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+public class CdaTemplateTest {
+
+    @Test
+    public void shouldRemoveCommentsFromXml() {
+
```
Comment (kmadej) [inline ihe-interop/src/test/java/org/motechproject/ihe/interop/domain/CdaTemplateTest.java]: empty line

**Comment by kmadej** [comment]
test this please

**Comment by kmadej** [comment]
test this please

**Pull request merged**